chef/cookbooks/cinder/recipes/ceph.rb

```diff
@@ -17,52 +17,75 @@
 # Recipe:: ceph
 #
 
-glance_servers = search(:node, "roles:glance-server")
-if glance_servers.length > 0
-  glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
-else   
-  glance_pool = nil
-end
+ceph_clients = {}
+ceph_keyrings = {}
 
 node[:cinder][:volumes].each_with_index do |volume, volid|
-  if volume['backend_driver'] == "rbd"
+  unless volume[:backend_driver] == "rbd"
+    next
+  else
+
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
+
+    cmd = ["ceph", "-k", admin_keyring, "-c", ceph_conf]
+    check_ceph = Mixlib::ShellOut.new(cmd)
 
-    check_ceph = Mixlib::ShellOut.new("ceph -k #{volume['rbd']['admin_keyring']} -c #{volume['rbd']['config_file']} -s | grep -q -e 'HEALTH_[OK|WARN]'")
-    check_ceph.run_command
+    if check_ceph.run_command.stdout.match("(HEALTH_OK|HEALTH_WARN)")
 
-    if check_ceph.exitstatus == 0
+      backend_id = "backend-#{volume[:backend_driver]}-#{volid}"
 
-      backend_id = "backend-#{volume['backend_driver']}-#{volid}"
       cinder_user = volume[:rbd][:user]
       cinder_pool = volume[:rbd][:pool]
 
-      cinder_pools = []
-      cinder_pools << cinder_pool
+      ceph_clients[ceph_conf] = {} unless ceph_clients[ceph_conf]
+      ceph_keyrings[ceph_conf] = admin_keyring unless ceph_keyrings[ceph_conf]
+
+      cinder_pools = (ceph_clients[ceph_conf][cinder_user] || []) << cinder_pool
+      ceph_clients[ceph_conf][cinder_user] = cinder_pools
+
+      ceph_pool cinder_pool do
+        ceph_conf ceph_conf
+        admin_keyring admin_keyring
+      end
+    else
+      Chef::Log.info("Ceph cluster is not ready yet.")
+    end
+  end
+end
+
+unless ceph_clients.empty?
+  glance_servers = search(:node, "roles:glance-server")
+  if glance_servers.length > 0
+    glance_pool = glance_servers[0][:glance][:rbd][:store_pool]
+  else
+    glance_pool = nil
+  end
+
+  ceph_clients.each do |ceph_conf, ceph_pools|
+    ceph_hash = Digest::MD5.hexdigest(ceph_conf)
+
+    ceph_pools.each_pair do |cinder_user, cinder_pools|
       cinder_pools << glance_pool unless glance_pool.nil?
+  
       allow_pools = cinder_pools.map{|p| "allow rwx pool=#{p}"}.join(", ")
       ceph_caps = { 'mon' => 'allow r', 'osd' => "allow class-read object_prefix rbd_children, #{allow_pools}" }
 
+      # We have to compute MD5 hash from ceph_conf and cinder_user 
+      # to be sure that keyring file will not be overwritten
+      ceph_hash = Digest::MD5.hexdigest(ceph_conf)
```

Comment:
One thing I don't like about this is that the username that are created in the ceph cluster will look really ugly now: md5-hash-of-config-file-name.username

See: https://github.com/crowbar/barclamp-cinder/pull/246/files#diff-4e35a687f5956e18c68d43a410d78e78R1947

It would be nice if we could avoid that. (Though I don't yet have an idea how to do that)